pkg/download/stasher/stasher.go

```diff
@@ -0,0 +1,132 @@
+package stasher
+
+import (
+	"context"
+	"io"
+	"time"
+
+	"github.com/gomods/athens/pkg/download"
+	"github.com/gomods/athens/pkg/errors"
+	"github.com/gomods/athens/pkg/stash"
+	"github.com/gomods/athens/pkg/storage"
+)
+
+type protocol struct {
+	s       storage.Backend
+	dp      download.Protocol
+	stasher stash.Stasher
+}
+
+// New takes an upstream Protocol and storage
+// it always prefers storage, otherwise it goes to upstream
+// and stashes the storage with the results through the given stasher.
+func New(dp download.Protocol, s storage.Backend, stasher stash.Stasher) download.Protocol {
+	p := &protocol{dp: dp, s: s, stasher: stasher}
+
+	return p
+}
+
+func (p *protocol) List(ctx context.Context, mod string) ([]string, error) {
+	return p.dp.List(ctx, mod)
+}
+
+func (p *protocol) Info(ctx context.Context, mod, ver string) ([]byte, error) {
+	const op errors.Op = "stasher.Info"
+	info, err := p.s.Info(ctx, mod, ver)
+	if errors.IsNotFoundErr(err) {
+		err = p.stasher.Stash(mod, ver)
+		if err != nil {
+			return nil, errors.E(op, err)
+		}
+		info, err = p.s.Info(ctx, mod, ver)
+	}
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+
+	return info, nil
+}
+
+func (p *protocol) Stash(mod, ver string) error {
```

Comment:
Why do we need this func?

Reply:
@marpio good catch, we don't need it. I originally made it like that but it made wrapping it with "pooling" and "singleflight" harder. I'll remove it.